File: ApiApplicationCpp/FullWorkflow3DProject/main/Helpers/ModelHelper.cpp
#include "pch.h"
#include <string>
#include <comutil.h>
#include "ModelHelper.h"
#include "../Utils.h"

#include <iostream>
#include <filesystem>

namespace fs = std::filesystem;

/// <summary>
/// Prepare the model: import file
/// </summary>
HRESULT ModelHelper::PrepareModel(ICamApiApplication* application,
    ICamApiProject* project,
    ICamApiPaths* pathsHelper) {
    const std::wstring importFilePath = L"Milling_25D\\Part1.igs";

    TResultStatus resultStatus = {};
    HRESULT result = S_OK;
    BSTR modelsFolder = nullptr;
    ICAMAPIGeometryImporter* importer = nullptr;

    try {
        // switch to model tab
        HRESULT hr = application->put_MainWorkMode(mwmModel);
        if (FAILED(hr))
            return hr;

        // get the path to the file we will import
        hr = pathsHelper->get_ModelsFolder(&modelsFolder);
        if (FAILED(hr) || !modelsFolder)
            throw std::runtime_error("get_ModelsFolder failed.");
        fs::path importFile = fs::path(modelsFolder) / importFilePath;
        std::cout << "Import file: " << importFile << "\n";
        if (!fs::exists(importFile))
            throw std::runtime_error("Cannot find file to import: " + importFile.string());

        // get importer
        hr = project->get_GeomImporter(&importer);
        if (FAILED(hr) || !importer)
            throw std::runtime_error("get_GeomImporter failed.");

        // import the file
        const _bstr_t bstrImportFile(importFile.native().c_str());
        const _bstr_t targetFolder(L"Part");
        hr = importer->ImportFile(bstrImportFile, targetFolder, false, &resultStatus);
        if (resultStatus.Code == rsError)
            throw std::runtime_error("ImportFile failed: " + Utils::BSTRToString(resultStatus.Description));
        if (FAILED(hr))
            throw std::runtime_error("ImportFile failed.");

        std::cout << "Model import completed." << std::endl;
    }
    catch (const std::exception& ex) {
        std::cerr << "Error in PrepareModel(): " << ex.what() << std::endl;
        result = E_FAIL;
    }

    // clean
    if (importer)
        importer->Release();
    if (modelsFolder)
        SysFreeString(modelsFolder);

    // return
    return result;
}

File: ApiApplicationNet/FullWorkflow3DProject/project/main/Helpers/ModelHelper.cs
using System;
using System.IO;
using CAMAPI.Application;
using CAMAPI.ApplicationMainForm;
using CAMAPI.DotnetHelper;
using CAMAPI.Project;
using CAMAPI.Singletons;

namespace ApplicationFullWorkflow3DProjectNet;

/// <summary>
/// Operations over the model
/// </summary>
public static class ModelHelper
{
    /// <summary>
    /// Relative path to the importing file
    /// </summary>
    private const string ImportFilePath = @"Milling_25D\Part1.igs";

    /// <summary>
    /// Prepare the model: import file
    /// </summary>
    public static void PrepareModel(ComWrapper<ICamApiApplication> applicationCom,
        ComWrapper<ICamApiProject> activeProjectCom,
        ComWrapper<ICamApiPaths> pathsHelperCom)
    {
        // switch to model tab
        applicationCom.Invoke(application =>
        {
            application.MainWorkMode = TMainWorkMode.mwmModel;
        });
        
        // import file
        ImportFile(activeProjectCom, pathsHelperCom);
    }

    /// <summary>
    /// Import a file to the project
    /// </summary>
    private static void ImportFile(ComWrapper<ICamApiProject> projectCom, ComWrapper<ICamApiPaths> pathsHelperCom)
    {
        // get the path to the file we will import
        var modelsFolder = pathsHelperCom.Invoke(pathsHelper => pathsHelper.ModelsFolder)
                           ?? throw new Exception("Cannot get models folder");
        var importFile = Path.Combine(modelsFolder, ImportFilePath);
        if (!File.Exists(importFile))
            throw new Exception($"Cannot find file to import: {importFile}");
        
        // import the file
        using var geomImporterCom = projectCom.InvokeAndWrap(project => project.GeomImporter);
        geomImporterCom.Invoke(importer =>
        {
            importer.ImportFile(importFile, "Part", false);
        });
    }
}